### PR TITLE
ToQueryString for ReadItemQueryingEnumerable

### LINKED ISSUE #20653: Cosmos: ToQueryString for ReadItemQueryingEnumerable

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #20653

### CODE REVIEW

File: src/EFCore.Cosmos/Properties/CosmosStrings.resx [hunk 1]
```diff
@@ -120,6 +120,9 @@
   <data name="CosmosNotInUse" xml:space="preserve">
     <value>Cosmos-specific methods can only be used when the context is using the Cosmos provider.</value>
   </data>
+  <data name="NoReadItemQueryString" xml:space="preserve">
+    <value>There is no string-based representation of this query.</value>
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Properties/CosmosStrings.resx:124]:
```suggestion
    <value>There is no string-based representation of this query as it's executed using 'ReadItemQueryAsync({resourceId}, {partitionKey})'.</value>
```

File: src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs [hunk 2]
```diff
@@ -61,9 +59,7 @@ public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToke
             IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
             public string ToQueryString()
-            {
-                throw new NotImplementedException("Cosmos: ToQueryString for ReadItemQueryingEnumerable #20653");
-            }
+                => CosmosStrings.NoReadItemQueryString;
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:62]:
```suggestion
               {
                    TryGetResourceId(out var resourceId);
                    TryGetPartitionId(out var partitionKey);
                    return CosmosStrings.NoReadItemQueryString(resourceId, partitionKey);
               }
```

Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Query/Internal/CosmosShapedQueryCompilingExpressionVisitor.ReadItemQueryingEnumerable.cs:62]: And move the implementation to the nested class

**Comment by ajcvickers** [comment]
@dotnet/efcore Also added split-query ToQueryString message addendum, as discussed in triage.

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs [hunk 3]
```diff
@@ -146,6 +148,18 @@ public override async Task Include_reference_and_collection(bool async)
 LEFT JOIN [Customers] AS [c] ON [o].[CustomerID] = [c].[CustomerID]
 LEFT JOIN [Order Details] AS [o0] ON [o].[OrderID] = [o0].[OrderID]
 ORDER BY [o].[OrderID], [c].[CustomerID], [o0].[OrderID], [o0].[ProductID]");
+
+            using var context = CreateContext();
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:152]: Please put this in it's own separate test.

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindSplitIncludeQuerySqlServerTest.cs [hunk 4]
```diff
@@ -180,6 +180,19 @@ public override async Task Include_reference_and_collection(bool async)
 LEFT JOIN [Customers] AS [c] ON [o].[CustomerID] = [c].[CustomerID]
 INNER JOIN [Order Details] AS [o0] ON [o].[OrderID] = [o0].[OrderID]
 ORDER BY [o].[OrderID], [c].[CustomerID]");
+
+            using var context = CreateContext();
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/NorthwindSplitIncludeQuerySqlServerTest.cs:184]: Like above.

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs (hunk 3 above)
Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:152]: Why this code still existing in different test? 😛

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindSplitIncludeQuerySqlServerTest.cs [hunk 5]
```diff
@@ -1759,6 +1778,16 @@ public override async Task Outer_idenfier_correctly_determined_when_doing_includ
 ORDER BY [c].[CustomerID], [o].[OrderID]");
         }
 
+        public override Task Include_collection_with_last_no_orderby(bool async)
+        {
+            return AssertTranslationFailedWithDetails(
+                () => AssertLast(
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/NorthwindSplitIncludeQuerySqlServerTest.cs:1784]: This should call base.

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs (hunk 3 above)
Comment (ajcvickers) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:152]: I moved it to a separate test, as requested.

Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:152]: But you did not remove the code from existing test. I will clean it up in one of the PRs.

**Pull request merged**